solutions/gold/ac-swap.mdx

```diff
@@ -0,0 +1,177 @@
+---
+id: ac-swap
+source: AC
+title: Swap
+author: Ryan Chou
+---
+
+<Spoiler title="Hint 1">
+
+$K \leq 10^9$ looks super large and unmanageable, but since $|S| \leq 30$, we should be able
+to construct a more reasonable upper bound.
+
+What's the maximum number of swaps such that we can get to every single possible string?
+
+</Spoiler>
+
+<Spoiler title="Answer to hint 1">
+
+In the worst case, we want to swap each letter from the back to the front of the string.
+Thus, $K \leq 30+29+28+\dots = \frac{30 \cdot 31}{2} = 465$.
+
+</Spoiler>
+
+<Spoiler title="Solution">
+
+[Official Analysis (Python)](https://atcoder.jp/contests/abc227/editorial/2920)
+
+## Explanation
+
+Instead of calculating swaps from an end state, let's compute our answer by
+adding one letter at a time to our final string. This enables us to compute
+transitions in $\mathcal{O}(1)$ by swapping in the next letter of some type.
+
+### State
+
+Our state will be $dp_{\texttt{cost}, K, E, Y} =$
+the number of ways to create a string with $\texttt{nK}$ $K$s, $\texttt{nE}$ $E$s, and
+$\texttt{nY}$ $Y$s, given that we've performed $\texttt{cost}$ swaps.
+
+### Base Case
+
+There is one way to make an empty string, so $dp_{0,0,0,0} = 1$.
+
+### Transitions
+
+Without loss of generality, let's assume that we want to place a $K$ in our string next.
+
+Then,
+
+$$
+dp[\texttt{cost} + \texttt{between}, \texttt{nK} + 1, \texttt{nE}, \texttt{nY}] \mathrel{+}= dp[\texttt{cost},\texttt{nK},\texttt{nE},\texttt{nY}]
+$$
+
+where $\texttt{between}$ are the amount of characters between our current position and the next
+occurence of $K$.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(S^3)$<Asterisk>Note that there is an additional constant
```

Comment:
```suggestion
**Time Complexity:** $\mathcal{O}(|S|^3)$<Asterisk>Note that there is an additional constant
```

Also, I would consider 465 as |S| choose 2 = O(|S|^2) instead of a constant factor but it's up to you

Reply:
great idea!